qutebrowser/browser/qutescheme.py

```diff
@@ -185,82 +185,17 @@ def qute_bookmarks(_url):
-        """
-        history = objreg.get('web-history').history_dict.values()
-        if reverse:
-            history = reversed(history)
-
-        # when history_dict is not reversed, we need to keep track of last item
-        # so that we can yield its atime
-        last_item = None
-
-        # end is 24hrs earlier than start
-        end_time = start_time - 24*60*60
-
-        for item in history:
-            # Skip redirects
-            # Skip qute:// links
-            if item.redirect or item.url.scheme() == 'qute':
-                continue
-
-            # Skip items out of time window
-            item_newer = item.atime > start_time
-            item_older = item.atime <= end_time
-            if reverse:
-                # history_dict is reversed, we are going back in history.
-                # so:
-                #     abort if item is older than start_time+24hr
-                #     skip if item is newer than start
-                if item_older:
-                    yield {"next": int(item.atime)}
-                    return
-                if item_newer:
-                    continue
-            else:
-                # history_dict isn't reversed, we are going forward in history.
-                # so:
-                #     abort if item is newer than start_time
-                #     skip if item is older than start_time+24hrs
-                if item_older:
-                    last_item = item
-                    continue
-                if item_newer:
-                    yield {"next": int(last_item.atime if last_item else -1)}
-                    return
-
-            # Use item's url as title if there's no title.
-            item_url = item.url.toDisplayString()
-            item_title = item.title if item.title else item_url
-            item_time = int(item.atime * 1000)
-
-            yield {"url": item_url, "title": item_title, "time": item_time}
-
-        # if we reached here, we had reached the end of history
-        yield {"next": int(last_item.atime if last_item else -1)}
-
-    if sys.hexversion >= 0x03050000:
-        # On Python >= 3.5 we can reverse the ordereddict in-place and thus
-        # apply an additional performance improvement in history_iter.
-        # On my machine, this gets us down from 550ms to 72us with 500k old
-        # items.
-        history = history_iter(start_time, reverse=True)
-    else:
-        # On Python 3.4, we can't do that, so we'd need to copy the entire
-        # history to a list. There, filter first and then reverse it here.
-        history = reversed(list(history_iter(start_time, reverse=False)))
-
-    return list(history)
+    # end is 24hrs earlier than start
+    end_time = start_time - 24*60*60
+    entries = objreg.get('web-history').entries_between(end_time, start_time)
+    return [{"url": e.url, "title": e.title or e.url, "time": e.atime * 1000}
+            for e in entries]
```

Comment:
The `WHERE atime >= start_time LIMIT 1000` query wouldn't work perfectly, because `atime` is not a unique key of `history`. So if there are multiple items with exactly the same timestamp on which the query splits, you would get duplicates in `qute:history`.

You could either make `atime` unique, which is probably too limiting, or add another unique key and use it in Javascript, or use this solution: Javascript could send the `atime` of the newest item shown in the history page along with the `OFFSET` value and the split query would be `WHERE ... and atime <= 'the-newest-timestamp' ORDER BY ... LIMIT 1000 OFFSET ...`. While there is no explicit protection from duplicates, the only way to get duplicates is that a new history item with exactly the same timestamp as the first item in the list appears - I can't imagine how that could happen if the time is monotonic.